mirrord/agent/src/steal/connection.rs

```diff
@@ -591,28 +604,131 @@ impl TcpConnectionStealer {
+
+        bytes.to_vec()
+    }
+
+    fn handle_response_reconstruction_fail(err: hyper::http::Error) -> hyper::http::Error {
+        error!("Could not reconstruct http response: {err:?}");
+        // TODO: What should happen? Currently just skipping that response, which would
+        //       also spoils the rest of the stream. Should hopefully never happen.
+        debug_assert!(false);
+        err
+    }
+
+    /// Forward an HTTP response to a stolen HTTP request from the layer to the back to the browser.
     ///
-    ///                       _________________________agent_________________________
-    /// Local App -> Layer -> ClientConnectionHandler -> Stealer -> HttpFilterManager -> Browser
-    ///                                                          ^- You are here.
-    async fn http_response(
-        &mut self,
-        response: HttpResponse,
-    ) -> std::result::Result<(), AgentError> {
-        todo!();
-        match self.port_subscriptions.get(&response.port) {
-            Some(HttpFiltered(manager)) => todo!(),
-            Some(Unfiltered(_)) | None => {
-                warn!(
-                    "Got an http response to port {:?} which is not being filtered. \
-                Maybe the layer unsubscribed and this is a remainder of a closed connection. \
-                Not forwarding.",
-                    response.port
-                );
-                Ok(())
+    ///                         _______________agent_______________
+    /// Local App --> Layer --> ClientConnectionHandler --> Stealer --> Browser
+    ///                                                             ^- You are here.
+    async fn http_response(&mut self, response: HttpResponse) -> Result<()> {
+        let tcp_stream_write_half;
+        if let Some(stream) = self.http_write_streams.get_mut(&response.connection_id) {
+            tcp_stream_write_half = stream;
+        } else {
+            warn!(
+                "Got an http response in a connection for which no tcp stream is present. Not forwarding.",
+            );
+            return Ok(());
+        }
+        let counter = self
+            .http_request_counters
+            .entry(response.connection_id)
+            .or_insert(0);
+        if response.request_id == *counter {
+            if let Ok(response) = response
+                .response
+                .try_into()
+                .map_err(Self::handle_response_reconstruction_fail)
+            {
+                tcp_stream_write_half
+                    .write_all(&Self::response_to_bytes(response).await)
+                    .await?; // TODO: handle error?
+            }
+            *counter += 1;
+            if let Some(queue) = self.http_response_queues.get_mut(&response.connection_id) {
+                while queue
+                    .peek()
+                    .is_some_and(|response| response.request_id == *counter)
```

Comment:
Is this supposed to be == and not <=?

Reply:
This is a good point, we should make sure it can't be `<`. `counter` holds the `request_id` of the first response not yet sent, so there should never be a response with a `request_id` lower than `counter`. If that holds (and we should make sure it does), then we'll only have `<=` when it's `==`, so I would write `==` so that it's clear that it cannot be `<`.
Added https://github.com/t4lz/mirrord/pull/3/commits/827c2bb753f494a2d9825456ba982ecaec15e728 to handle the case where a client (layer) sends the stealer a response with a duplicate `request_id` and make sure we do not insert a response with a ` request_id` lower than the corresponding counter to the corresponding queue.

Reply:
Oh okay. I misunderstood anyway, didn't realize it was the request_id, not the response id.